arangod/RocksDBEngine/RocksDBMetaCollection.cpp

```diff
@@ -527,56 +524,84 @@ rocksdb::SequenceNumber RocksDBMetaCollection::serializeRevisionTree(
+      newTree->insert(revisions);
+      revisions.clear();
+    
+      if (_logicalCollection.vocbase().server().isStopping()) {
+        THROW_ARANGO_EXCEPTION(TRI_ERROR_SHUTTING_DOWN);
+      }
+    }
+    it.next();
+  }
+  if (!revisions.empty()) {
+    newTree->insert(revisions);
+  }
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  newTree->checkConsistency();
+#endif
+
+  return std::make_pair(std::move(newTree), state->beginSeq());
+}
+
 Result RocksDBMetaCollection::rebuildRevisionTree() {
   return basics::catchToResult([this]() -> Result {
     std::unique_lock<std::mutex> guard(_revisionTreeLock);
-    auto ctxt = transaction::StandaloneContext::Create(_logicalCollection.vocbase());
-    SingleCollectionTransaction trx(ctxt, _logicalCollection, AccessMode::Type::READ);
-    Result res = trx.begin();
-    if (res.fail()) {
-      LOG_TOPIC("d1e53", WARN, arangodb::Logger::ENGINES)
-          << "failed to begin transaction to rebuild revision tree "
-             "for collection '"
-          << _logicalCollection.id().id() << "'";
-      return res;
-    }
-    auto* state = RocksDBTransactionState::toState(&trx);
-
-    std::vector<std::uint64_t> revisions;
-    auto iter = getReplicationIterator(ReplicationIterator::Ordering::Revision, trx);
-    if (!iter) {
-      LOG_TOPIC("d1e54", WARN, arangodb::Logger::ENGINES)
-          << "failed to retrieve replication iterator to rebuild revision tree "
-             "for collection '"
-          << _logicalCollection.id().id() << "'";
-      return Result(TRI_ERROR_INTERNAL);
-    }
-    RevisionReplicationIterator& it =
-        *static_cast<RevisionReplicationIterator*>(iter.get());
-    auto newTree = allocateEmptyRevisionTree();
-    while (it.hasMore()) {
-      revisions.emplace_back(it.revision().id());
-      if (revisions.size() >= 5000) {  // arbitrary batch size
-        newTree->insert(revisions);
-        revisions.clear();
-      }
-      it.next();
-    }
-    if (!revisions.empty()) {
-      newTree->insert(revisions);
+    
+    auto result = revisionTreeFromCollection();
+    if (result.fail()) {
+      return result.result(); 
     }
-    _revisionTree = std::move(newTree);
-    _revisionTreeApplied = state->beginSeq();
-    _revisionTreeCreationSeq = state->beginSeq();
-    _revisionTreeSerializedSeq = state->beginSeq();
-    return Result();
+
+    auto&& [newTree, beginSeq] = result.get();
+
+    _revisionTree = std::make_unique<RevisionTreeAccessor>(std::move(newTree), _logicalCollection);
+    _revisionTreeApplied = beginSeq;
+    _revisionTreeCreationSeq = beginSeq;
+    _revisionTreeSerializedSeq = beginSeq;
+  
+    return {};
   });
 }
 
 void RocksDBMetaCollection::rebuildRevisionTree(std::unique_ptr<rocksdb::Iterator>& iter) {
-  std::unique_lock<std::mutex> guard(_revisionTreeLock);
+  auto newTree = allocateEmptyRevisionTree(revisionTreeDepth);
 
-  _revisionTree = allocateEmptyRevisionTree();
+  std::unique_lock<std::mutex> guard(_revisionTreeLock);
```

Comment:
Wouldn't it be sufficient to acquire the lock later once we actually update our members?

Reply:
Will be incorporated into https://github.com/arangodb/arangodb/pull/14169